src/abe/policy.c

```diff
@@ -0,0 +1,350 @@
+            if (c1 == -1) {
+                return -1;
+            }
+            c_out = boolean_to_msp_iterative(&msp2, bool_exp2, &vec2, (size_t) c1);
+            if (c_out == -1) {
+                return -1;
+            }
+            found = true;
+            break;
+        }
+        if (num_brc == 0 && i < strlen(bool_exp) - 2 && bool_exp[i] == 'O' &&
+                bool_exp[i+1] == 'R') {
+            bool_exp1 = substring(bool_exp, 0, i);
+            bool_exp2 = substring(bool_exp, i + 2, strlen(bool_exp));
+            c1 = boolean_to_msp_iterative(&msp1, bool_exp1, vec, c);
+            if (c1 == -1) {
+                return -1;
+            }
+            c_out = boolean_to_msp_iterative(&msp2, bool_exp2, vec, (size_t) c1);
+            if (c_out == -1) {
+                return -1;
+            }
+            found = true;
+            break;
+        }
+
+    }
+    if (found == false) {
+        if (bool_exp[0] == '(' && bool_exp[strlen(bool_exp) - 1] == ')') {
+            bool_exp = substring(bool_exp, 1, strlen(bool_exp) - 1);
+            return boolean_to_msp_iterative(msp, bool_exp, vec, c);
+        }
+
+        int attrib = str_to_int(bool_exp);
+        if (attrib == -1) {
+            return -1;
+        }
+
+        msp->mat = (cfe_mat*) cfe_malloc(sizeof(cfe_mat));
+        cfe_mat_init(msp->mat, 1, (size_t) c);
+        mpz_t zero;
+        mpz_init_set_ui(zero, 0);
+        for (size_t i = 0; i < c; i++) {
+            if (i < vec->size) {
+                cfe_mat_set(msp->mat, vec->vec[i], 0, i);
+            } else {
+                cfe_mat_set(msp->mat, zero, 0, i);
+            }
+        }
+
+        msp->row_to_attrib = (int*) cfe_malloc(sizeof(int) * 1);
+        msp->row_to_attrib[0] = attrib;
+        return (int) c;
+    } else {
+        msp->mat = (cfe_mat*) cfe_malloc(sizeof(cfe_mat));
+        msp->row_to_attrib = (int*) cfe_malloc(sizeof(int) * (msp1.mat->rows + msp2.mat->rows));
+        cfe_mat_init(msp->mat, msp1.mat->rows + msp2.mat->rows, (size_t) c_out);
+        mpz_t tmp;
+        mpz_init(tmp);
+        for (size_t i = 0; i < msp1.mat->rows; i++) {
+            for (size_t j = 0; j < msp1.mat->cols; j++) {
+                cfe_mat_get(tmp, msp1.mat, i, j);
+                cfe_mat_set(msp->mat, tmp, i, j);
+            }
+            mpz_set_ui(tmp, 0);
+            for (size_t j = msp->mat->cols; j < (size_t) c_out; j++) {
+                cfe_mat_set(msp->mat, tmp, i, j);
+            }
+            msp->row_to_attrib[i] = msp1.row_to_attrib[i];
+        }
+        for (size_t i = 0; i < msp2.mat->rows; i++) {
+            for (size_t j = 0; j < (size_t) c_out; j++) {
+                cfe_mat_get(tmp, msp2.mat, i, j);
+                cfe_mat_set(msp->mat, tmp, i + msp1.mat->rows, j);
+            }
+            msp->row_to_attrib[i + msp1.mat->rows] = msp2.row_to_attrib[i];
+        }
+        return c_out;
+    }
+}
```

Comment:
This function leaks memory everywhere. Make sure each memory initialization has a corresponding statement that frees it somewhere.